devito/types/dimension.py

```diff
@@ -1440,8 +1469,112 @@ def _arg_values(self, *args, **kwargs):
         return values
 
 
-# ***
-# Utils
+# *** Utils
+
+
+class IndexAccessFunction(sympy.Add):
+
+    """
+    A IndexAccessFunction is an expression used to index into a Function.
+    """
+
+    # Prioritize self's __add__ and __sub__ to construct AffineIndexAccessFunction
+    _op_priority = sympy.Add._op_priority + 1.
+
+    def __eq__(self, other):
+        return super().__eq__(other)
+
+    __hash__ = sympy.Add.__hash__
+
+
+class AffineIndexAccessFunction(IndexAccessFunction):
+    """
+    An AffineIndexAccessFunction is the sum of three operands:
+
+        * the "main" Dimension (in practice a SpaceDimension or a TimeDimension);
+        * an offset (number or symbolic expression, of dtype integer).
+        * a StencilDimension;
+
+    Examples
+    --------
+    The AffineIndexAccessFunction `x + sd + 3`, with `sd \in [-2, 2]`, represents
+    the index access functions `[x + 1, x + 2, x + 3, x + 4, x + 5]`
+    """
+
+    def __new__(cls, *args, **kwargs):
+        d = 0
+        sd = 0
+        ofs_items = []
+        for a in args:
+            if isinstance(a, StencilDimension):
+                if sd != 0:
+                    return sympy.Add(*args, **kwargs)
+                sd = a
+            elif isinstance(a, Dimension):
+                d = cls._separate_dims(d, a, ofs_items)
+                if d is None:
+                    return sympy.Add(*args, **kwargs)
+            elif isinstance(a, AffineIndexAccessFunction):
+                if sd != 0 and a.sd != 0:
+                    return sympy.Add(*args, **kwargs)
+                d = cls._separate_dims(d, a.d, ofs_items)
+                if d is None:
+                    return sympy.Add(*args, **kwargs)
+                sd = a.sd
+                ofs_items.append(a.ofs)
+            else:
+                ofs_items.append(a)
+
+        ofs = sympy.Add(*[i for i in ofs_items if i is not None])
+        if not all(is_integer(i) or i.is_Symbol for i in ofs.free_symbols):
+            return sympy.Add(*args, **kwargs)
+
+        obj = IndexAccessFunction.__new__(cls, d, ofs, sd)
```

Comment:
That's a bit convoluted no? This will redo all the if/else. Shouldn't this just be `obj = sympy.Add.__new__(cls, d, ofs, sd)`

Reply:
I actually have a hard time figuring out why this doesn't overflow I must be missing something

Reply:
No, it's OK, it's calling the `IndexAccessFunction`'s `__new__`. But `IndexAccessFunction` doesn't have a `__new__` (yet...), so it falls back to `sympy.Add.__new__` since `sympy.Add` is `IndexAccessFunction`'s base class.